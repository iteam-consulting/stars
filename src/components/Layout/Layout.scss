@import '../variables';

.flex {
  display: flex;

  &.x {
    flex-direction: row;
  }

  &.y {
    flex-direction: column;
  }

  &.-x {
    flex-direction: row-reverse;
  }

  &.-y {
    flex-direction: column-reverse;
  }
}

.sticky {
  position: sticky;
  top: 0;
}

.fixed {
  position: fixed;
}

.absolute {
  position: absolute;
}

.pinned {
  &:not(.center):not(.start):not(.end) {
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
  }

  &.center {
    top: 0;
    bottom: 0;
  }

  &.start {
    top: 0;
    bottom: 0;
    left: 0;
  }

  &.end {
    top: 0;
    right: 0;
    bottom: 0;
  }
}

.fill {
  &.flex {
    flex: 1;
  }

  &:not(.flex) {
    height: 100%;
    width: 100%;
  }
}

.center {
  &.flex {
    margin: 0 auto;
  }

  &:not(.flex) {
    left: 50%;
    transform: translateX(-50%);
  }
}

.start {
  &.flex {
    align-self: flex-start;
  }

  &:not(.flex) {
    left: 0;
  }
}

.end {
  &.flex {
    align-self: flex-end;
  }

  &:not(.flex) {
    margin-left: auto;
    right: 0;
  }
}

@each $key, $val in $spacing {
  .spacing-#{$key} {
    padding: $val;
  }
}
