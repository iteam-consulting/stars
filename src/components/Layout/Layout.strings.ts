export class LayoutStrings {
  static InvalidPinningProps: string =
    'Layout with pinned will override any top|right|bottom|left position prop.';
}
