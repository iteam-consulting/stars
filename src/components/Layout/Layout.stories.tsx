import { Meta, Story } from '@storybook/react';
import Layout, { LayoutProps } from './Layout';

export default {
  title: 'Components/Layout',
  component: Layout,
} as Meta;

const Template: Story<LayoutProps> = args => <Layout {...args} />;

export const Default = Template.bind({});
Default.args = {
  children: 'Layout',
  style: {
    backgroundColor: 'cornflowerblue',
  },
};

export const FixedWidth = Template.bind({});
FixedWidth.args = {
  children: '250px',
  style: {
    width: 250,
    backgroundColor: 'cornflowerblue',
  },
  mode: 'absolute',
};
