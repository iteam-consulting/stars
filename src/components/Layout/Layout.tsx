import classNames from 'classnames';
import { CSSProperties, forwardRef, ReactNode } from 'react';
import { HTMLAttributes } from 'react';
import { LayoutStrings } from './Layout.strings';

type LayoutModes = 'flex' | 'sticky' | 'fixed' | 'absolute';
type LayoutSpacing = 'xs' | 'sm' | 'md' | 'lg' | 'xl';

export type LayoutProps = {
  children?: ReactNode;
  mode?: LayoutModes;
  top?: number;
  right?: number;
  bottom?: number;
  left?: number;
  spacing?: LayoutSpacing;
  fill?: boolean;
  pinned?: boolean;
} & HTMLAttributes<HTMLDivElement>;

const Layout = forwardRef<HTMLDivElement, LayoutProps>(function Layout(
  props,
  ref
) {
  const {
    className,
    mode,
    pinned,
    fill,
    spacing,
    top,
    right,
    bottom,
    left,
    style,
    ...rest
  } = props;
  const classes = classNames(className, mode, {
    fill,
    pinned,
    [`spacing-${spacing}`]: spacing,
  });

  const styles: CSSProperties = {
    ...style,
    top: pinned ? 0 : top,
    right: pinned ? 0 : right,
    bottom: pinned ? 0 : bottom,
    left: pinned ? 0 : left,
  };

  if (pinned && (top || right || bottom || left)) {
    console.warn(LayoutStrings.InvalidPinningProps);
  }

  return <div ref={ref} {...rest} style={styles} className={classes} />;
});

export default Layout;
