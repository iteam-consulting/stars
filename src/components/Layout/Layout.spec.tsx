import { render } from '@testing-library/react';
import Layout from './Layout';
import { LayoutStrings } from './Layout.strings';

const warn = console.warn;

describe('Layout', () => {
  const warnMock = jest.fn();

  afterEach(() => {
    console.warn = warn;
    warnMock.mockReset();
  });

  beforeEach(() => {
    console.warn = warnMock;
  });

  test('Layout should render properly', () => {
    const { container } = render(<Layout />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('Layout should render classes based on props', () => {
    const { getByTestId } = render(
      <Layout data-testid="layout" mode="flex" left={50} spacing="xs" />
    );
    const element = getByTestId('layout');

    expect(element.classList).toContain('flex');
    expect(element.classList).toContain('spacing-xs');
    expectPositionStyle(element, 'left', 50);
  });

  test('Layout pinning should validate props', () => {
    const { getByTestId } = render(
      <Layout data-testid="layout" pinned mode="absolute" left={10} />
    );
    expect(warnMock).toHaveBeenCalledWith(LayoutStrings.InvalidPinningProps);

    const element = getByTestId('layout');

    expectPositionStyle(element, 'top', 0);
    expectPositionStyle(element, 'right', 0);
    expectPositionStyle(element, 'bottom', 0);
    expectPositionStyle(element, 'left', 0);
  });
});

function expectPositionStyle(
  element: HTMLElement,
  pos: 'top' | 'right' | 'bottom' | 'left',
  value: number
) {
  expect(element.style[pos]).toBe(`${value}px`);
}
