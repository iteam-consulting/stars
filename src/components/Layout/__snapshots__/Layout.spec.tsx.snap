// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout Layout should render properly 1`] = `
<div
  class=""
/>
`;
